test/lit/passes/optimize-instructions-gc.wast

```diff
@@ -90,4 +90,331 @@
+  ;; CHECK-NEXT:   (local.get $struct)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (local.get $func)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (local.get $data)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (local.get $i31)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $unneeded_as
+   (param $struct (ref $struct))
+   (param $func (ref func))
+   (param $data (ref data))
+   (param $i31 (ref i31))
+   (drop
+    (ref.as_non_null (local.get $struct))
+   )
+   (drop
+    (ref.as_func (local.get $func))
+   )
+   (drop
+    (ref.as_data (local.get $data))
+   )
+   (drop
+    (ref.as_i31 (local.get $i31))
+   )
+  )
+
+  ;; similar to $unneeded_as, but the values are nullable. we can turn the
+  ;; more specific things into ref.as_non_null.
+  ;; CHECK:      (func $unneeded_as_null (param $struct (ref null $struct)) (param $func funcref) (param $data (ref null data)) (param $i31 (ref null i31))
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (ref.as_non_null
+  ;; CHECK-NEXT:    (local.get $struct)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (ref.as_non_null
+  ;; CHECK-NEXT:    (local.get $func)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (ref.as_non_null
+  ;; CHECK-NEXT:    (local.get $data)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (ref.as_non_null
+  ;; CHECK-NEXT:    (local.get $i31)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $unneeded_as_null
+   (param $struct (ref null $struct))
+   (param $func (ref null func))
+   (param $data (ref null data))
+   (param $i31 (ref null i31))
+   (drop
+    (ref.as_non_null (local.get $struct))
+   )
+   (drop
+    (ref.as_func (local.get $func))
+   )
+   (drop
+    (ref.as_data (local.get $data))
+   )
+   (drop
+    (ref.as_i31 (local.get $i31))
+   )
+  )
+  ;; similar to $unneeded_as, but the values are of mixed kind (as_func of
+  ;; data, etc.), so we know we will trap
+  ;; CHECK:      (func $unneeded_as_bad_kinds (param $func funcref) (param $data (ref null data)) (param $i31 (ref null i31))
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (block
+  ;; CHECK-NEXT:    (drop
+  ;; CHECK-NEXT:     (local.get $data)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:    (unreachable)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (block
+  ;; CHECK-NEXT:    (drop
+  ;; CHECK-NEXT:     (local.get $i31)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:    (unreachable)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (block
+  ;; CHECK-NEXT:    (drop
+  ;; CHECK-NEXT:     (local.get $func)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:    (unreachable)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $unneeded_as_bad_kinds
+   (param $func (ref null func))
+   (param $data (ref null data))
```

Comment:
```suggestion
   (param $data (ref data))
```
Make one of the params non-nullable to test both nullable and non-nullable cases?

Reply:
Good catch. I'll add three more testcases below to check those, and a comment.